extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,779 @@
+        partition.getPartitionId(),
+        iteratorEnum.toString(),
+        sequenceNumber
+    ).getShardIterator();
+
+    checkPartitionsStarted = true;
+  }
+
+  private void filterBufferAndResetFetchRunnable(Set<StreamPartition<String>> partitions) throws InterruptedException
+  {
+    scheduledExec.shutdown();
+
+    try {
+      if (!scheduledExec.awaitTermination(EXCEPTION_RETRY_DELAY_MS, TimeUnit.MILLISECONDS)) {
+        scheduledExec.shutdownNow();
+      }
+    }
+    catch (InterruptedException e) {
+      log.warn(e, "InterruptedException while shutting down");
+      throw e;
+    }
+
+    scheduledExec = Executors.newScheduledThreadPool(
+        fetchThreads,
+        Execs.makeThreadFactory("KinesisRecordSupplier-Worker-%d")
+    );
+
+    // filter records in buffer and only retain ones whose partition was not seeked
+    BlockingQueue<OrderedPartitionableRecord<String, String>> newQ = new LinkedBlockingQueue<>(recordBufferSize);
+    records
+        .stream()
+        .filter(x -> !partitions.contains(x.getStreamPartition()))
+        .forEachOrdered(newQ::offer);
+
+    records = newQ;
+
+    // restart fetching threads
+    partitionResources.values().forEach(x -> x.started = false);
+    checkPartitionsStarted = true;
+  }
+
+  @Nullable
+  private String getSequenceNumberInternal(StreamPartition<String> partition, ShardIteratorType iteratorEnum)
+  {
+
+    String shardIterator = null;
+    try {
+      shardIterator = kinesis.getShardIterator(
+          partition.getStream(),
+          partition.getPartitionId(),
+          iteratorEnum.toString()
+      ).getShardIterator();
+    }
+    catch (ResourceNotFoundException e) {
+      log.warn(e, "Caught ResourceNotFoundException while getting shardIterator");
+    }
+
+    return getSequenceNumberInternal(partition, shardIterator);
+  }
+
+  @Nullable
+  private String getSequenceNumberInternal(StreamPartition<String> partition, String shardIterator)
+  {
+    long timeoutMillis = System.currentTimeMillis() + fetchSequenceNumberTimeout;
+
+    while (shardIterator != null && System.currentTimeMillis() < timeoutMillis) {
+
+      if (closed) {
+        log.info("KinesisRecordSupplier closed while fetching sequenceNumber");
+        return null;
+      }
+
+      GetRecordsResult recordsResult;
+      try {
+        // we call getRecords with limit 1000 to make sure that we can find the first (earliest) record in the shard.
+        // In the case where the shard is constantly removing records that are past their retention period, it is possible
+        // that we never find the first record in the shard if we use a limit of 1.
+        recordsResult = kinesis.getRecords(new GetRecordsRequest().withShardIterator(shardIterator).withLimit(1000));
+      }
+      catch (ProvisionedThroughputExceededException e) {
+        log.warn(
+            e,
+            "encounted ProvisionedThroughputExceededException while fetching records, this means "
```

Comment:
typo: encounted -> encountered

Also, you can recommend that they can increase the number of shards to get more throughput